File: backend/app/services/store.py
import uuid
from app.supabase_rest import select, insert
from app.services.email import clean_html, extract_preview
from app.realtime_bus import publish_event
from app.mail_crypto import encrypt_mail_json, encrypt_mail_text
from app.services.security import sanitize_headers
from app.services.cache import bump_user_cache_version


async def store_email(user_id: str, account_id: str, message: dict) -> str:
    email_id = str(uuid.uuid4())
    body_html = clean_html(message.get("body") or "")
    preview = message.get("snippet") or extract_preview(body_html)
    safe_headers = sanitize_headers(message.get("headers") or {})

    existing = await select(
        "emails",
        "id",
        filters=[("account_id", "eq", account_id), ("message_id", "eq", message.get("message_id"))],
        use_service=True,
    )
    if existing:
        return existing[0]["id"]

    await insert(
        "emails",
        {
            "id": email_id,
            "user_id": user_id,
            "account_id": account_id,
            "provider": message.get("provider"),
            "message_id": message.get("message_id"),
            "thread_id": message.get("thread_id"),
            "subject": message.get("subject"),
            "sender_name": message.get("sender_name"),
            "sender_email": message.get("sender_email"),
            "preview_snippet": preview,
            "body_html": None,
            "body_html_enc": encrypt_mail_text(body_html),
            "received_at": message.get("received_at").isoformat() if hasattr(message.get("received_at"), "isoformat") else message.get("received_at"),
            "raw_headers": None,
            "raw_headers_enc": encrypt_mail_json(safe_headers),
        },
        use_service=True,
    )
    await publish_event(
        user_id,
        {
            "type": "email_received",
            "email_id": email_id,
            "processing_status": "pending",
        },
    )
    await bump_user_cache_version(user_id)
    return email_id

File: backend2/src/services/store.ts
import { randomUUID } from "node:crypto";

import { encryptMailJson, encryptMailText } from "../mailCrypto.js";
import { publishEvent } from "../realtimeBus.js";
import { insert, select } from "../supabaseRest.js";
import { bumpUserCacheVersion } from "./cache.js";
import { cleanHtml, extractPreview } from "./email.js";
import { sanitizeHeaders } from "./security.js";

export async function storeEmail(userId: string, accountId: string, message: Record<string, any>): Promise<string> {
  const existing = await select("emails", "id", {
    filters: [
      ["account_id", "eq", accountId],
      ["message_id", "eq", String(message.message_id ?? "")]
    ],
    useService: true
  });
  if (existing[0]?.id) return existing[0].id;

  const bodyHtml = cleanHtml(String(message.body ?? ""));
  const preview = String(message.snippet ?? "") || extractPreview(bodyHtml);
  const safeHeaders = sanitizeHeaders(message.headers ?? {});
  const emailId = randomUUID();

  await insert(
    "emails",
    {
      id: emailId,
      user_id: userId,
      account_id: accountId,
      provider: message.provider,
      message_id: message.message_id,
      thread_id: message.thread_id,
      subject: message.subject,
      sender_name: message.sender_name,
      sender_email: message.sender_email,
      preview_snippet: preview,
      body_html: null,
      body_html_enc: encryptMailText(bodyHtml),
      received_at: message.received_at,
      raw_headers: null,
      raw_headers_enc: encryptMailJson(safeHeaders)
    },
    { useService: true }
  );

  await publishEvent(userId, { type: "email_received", email_id: emailId, processing_status: "pending" });
  await bumpUserCacheVersion(userId);
  return emailId;
}
